src/NewtonKrylov.fypp

```diff
@@ -0,0 +1,418 @@
+         solver => linear_solver
+      else
+         solver => gmres_${type[0]}$${kind}$
+      end if
+      ! Linear solver options ?
+      if (present(linear_solver_options)) then
+         has_solver_opts = .true.
+         allocate(solver_opts, source=linear_solver_options)
+      else
+         has_solver_opts = .false.
+      end if
+      ! Preconditioner ?
+      if (present(preconditioner)) then
+         has_precond = .true.
+         allocate(precond, source=preconditioner)
+      else
+         has_precond = .false.
+      end if
+      ! Scheduler
+      if (present(scheduler)) then
+         tolerance_scheduler => scheduler
+      else
+         tolerance_scheduler => constant_atol_${kind}$
+      endif
+
+      ! Initialisation      
+      maxiter = opts%maxiter ; maxstep_bisection = opts%maxstep_bisection ;
+      converged = .false.
+      allocate(residual, source=X); call residual%zero()
+      allocate(increment,source=X); call increment%zero()
+
+      call sys%eval(X, residual, target_tol)
+      rnorm = residual%norm()
+      ! Check for lucky convergence.
+      if (rnorm < target_tol) then
+         write(msg,'(A)') 'Initial guess is a fixed point to tolerance!'
+         call logger%log_warning(msg, module=this_module, procedure='newton_${type[0]}$${kind}$')
+         converged = .true.
+         return
+      end if
+
+      write(msg,'(A)') 'Starting Newton iteration ...'
+      call logger%log_information(msg, module=this_module, procedure='newton_${type[0]}$${kind}$')
+      ! Newton iteration
+      newton: do i = 1, maxiter
+
+         ! Set dynamic tolerances for Newton iteration and linear solves.
+         call tolerance_scheduler(tol, target_tol, rnorm, i, info)
+         write(msg,"(A,I0,3(A,E9.2))") 'Start step ', i, ': rnorm= ', rnorm, ', tol= ', tol, ', target= ', target_tol
+         call logger%log_message(msg, module=this_module, procedure='newton_${type[0]}$${kind}$')
+
+         ! Define the Jacobian
+         sys%jacobian%X = X
+        
+         ! Solve the linear system using GMRES.
+         call residual%chsgn()
+         if (.not. has_precond .and. .not. has_solver_opts) then
+            call solver(sys%jacobian, residual, increment, info, atol=tol, &
+                  &                                              transpose=.false.)
+         elseif (.not. has_precond .and. has_solver_opts) then
+            call solver(sys%jacobian, residual, increment, info, atol=tol,  &
+                  &                         options=solver_opts, transpose=.false.)
+         elseif (has_precond .and. .not. has_solver_opts) then
+            call solver(sys%jacobian, residual, increment, info, atol=tol, &
+                  & preconditioner=precond,                      transpose=.false.)
+         else
+            call solver(sys%jacobian, residual, increment, info, atol=tol, &
+                  & preconditioner=precond, options=solver_opts, transpose=.false.)
+         end if
```

Comment:
Could this be simply replaced by

```
call solver(sys%jacobian, residual, increment, info, atol=tol, transpose=.false., opts=linear_solver_opts, preconditioner=precond)
```

Since `linear_solver_opts` and `precond` are optional both here and in the `linear_solver` interface, checking whether these have been defined or not should be done in the linear solver itself, no here. Does that make sense ?

Reply:
Yes, that's possible. For the same reason I can also remove the presence checks for these options making the whole code more succinct.